# Who will accept the coupon?

## Overview

The goal of this project is to explore a dataset of driving scenarios where users were offered coupons and determine what types of customers are more likely to accept them. By analyzing customer demographics, behavior, and contextual factors, we aim to identify patterns that influence coupon acceptance.

## Context

Imagine driving through town and a coupon is delivered to your cell phone for a restaurant near where you are driving. Would you accept that coupon and take a short detour to the restaurant? Would you accept the coupon but use it on a subsequent trip? Would you ignore the coupon entirely? What if the coupon was for a bar instead of a restaurant? What about a coffee house? Would you accept a bar coupon with a minor passenger in the car? What about if it was just you and your partner in the car? Would weather impact the rate of acceptance? What about the time of day?

## Data

This data comes from the UCI Machine Learning repository and was collected via a survey on Amazon Mechanical Turk. The survey describes different driving scenarios including the destination, current time, weather, passenger, etc., and then ask the person whether he will accept the coupon if he is the driver. Answers that the user will drive there ‘right away’ or ‘later before the coupon expires’ are labeled as ‘Y = 1’ and answers ‘no, I do not want the coupon’ are labeled as ‘Y = 0’.  There are five different types of coupons -- less expensive restaurants (under \\$20), coffee houses, carry out & take away, bar, and more expensive restaurants (\\$20 - $50).

## Data Description

The attributes of this data set include:

1. User attributes
    -  Gender: male, female
    -  Age: below 21, 21 to 25, 26 to 30, etc.
    -  Marital Status: single, married partner, unmarried partner, or widowed
    -  Number of children: 0, 1, or more than 1
    -  Education: high school, bachelors degree, associates degree, or graduate degree
    -  Occupation: architecture & engineering, business & financial, etc.
    -  Annual income: less than \\$12500, \\$12500 - \\$24999, \\$25000 - \\$37499, etc.
    -  Number of times that he/she goes to a bar: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she buys takeaway food: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she goes to a coffee house: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she eats at a restaurant with average expense less than \\$20 per person: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    -  Number of times that he/she goes to a bar: 0, less than 1, 1 to 3, 4 to 8 or greater than 8
    
2. Contextual attributes
    - Driving destination: home, work, or no urgent destination
    - Location of user, coupon and destination: we provide a map to show the geographical location of the user, destination, and the venue, and we mark the distance between each two places with time of driving. The user can see whether the venue is in the same direction as the destination.
    - Weather: sunny, rainy, or snowy
    - Temperature: 30F, 55F, or 80F
    - Time: 10AM, 2PM, or 6PM
    - Passenger: alone, partner, kid(s), or friend(s)

3. Coupon attributes
    - time before it expires: 2 hours or one day

## Tools and Libraries

The project uses Python and several common data analysis libraries:

-   Python
-   Pandas
-   NumPy
-   Plotly
-   Seaborn / Matplotlib
-   Jupyter Notebook

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import plotly.express as px

In [27]:
original_data = pd.read_csv('data/coupons.csv')
data = original_data.copy()
print(f"Number of entries: {original_data.shape[0]}.")
original_data.head()

Number of entries: 12684.


,destination,passanger,weather,temperature,time,coupon,expiration,gender,age,maritalStatus,...,CoffeeHouse,CarryAway,RestaurantLessThan20,Restaurant20To50,toCoupon_GEQ5min,toCoupon_GEQ15min,toCoupon_GEQ25min,direction_same,direction_opp,Y
0,No Urgent Place,Alone,Sunny,55,2PM,Restaurant(<20),1d,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,0,0,0,1,1
1,No Urgent Place,Friend(s),Sunny,80,10AM,Coffee House,2h,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,0,0,0,1,0
2,No Urgent Place,Friend(s),Sunny,80,10AM,Carry out & Take away,2h,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,1,0,0,1,1
3,No Urgent Place,Friend(s),Sunny,80,2PM,Coffee House,2h,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,1,0,0,1,0
4,No Urgent Place,Friend(s),Sunny,80,2PM,Coffee House,1d,Female,21,Unmarried partner,...,never,NaN,4~8,1~3,1,1,0,0,1,0


## Missing and Problematic Data

First look for missing data.

In [14]:
data.isnull().sum()

destination                 0
passanger                   0
weather                     0
temperature                 0
time                        0
coupon                      0
expiration                  0
gender                      0
age                         0
maritalStatus               0
has_children                0
education                   0
occupation                  0
income                      0
car                     12576
Bar                       107
CoffeeHouse               217
CarryAway                 151
RestaurantLessThan20      130
Restaurant20To50          189
toCoupon_GEQ5min            0
toCoupon_GEQ15min           0
toCoupon_GEQ25min           0
direction_same              0
direction_opp               0
Y                           0
dtype: int64

In [15]:
print(original_data['car'].value_counts())

car
Scooter and motorcycle                      22
Mazda5                                      22
do not drive                                22
crossover                                   21
Car that is too old to install Onstar :D    21
Name: count, dtype: int64


There are 12684 entries, of which 12576 are missing data for the 'car' column. The data present for the car column is not very useful. This column will not be used and dropped.

In [31]:
data = data.drop(columns=['car'])
data.columns

Index(['destination', 'passanger', 'weather', 'temperature', 'time', 'coupon',
       'expiration', 'gender', 'age', 'maritalStatus', 'has_children',
       'education', 'occupation', 'income', 'Bar', 'CoffeeHouse', 'CarryAway',
       'RestaurantLessThan20', 'Restaurant20To50', 'toCoupon_GEQ5min',
       'toCoupon_GEQ15min', 'toCoupon_GEQ25min', 'direction_same',
       'direction_opp', 'Y'],
      dtype='object')

In [32]:
print(data['Bar'].value_counts())
print(data['CoffeeHouse'].value_counts())
print(data['CarryAway'].value_counts())
print(data['RestaurantLessThan20'].value_counts())
print(data['Restaurant20To50'].value_counts())

Bar
never    5304
less1    3482
1~3      2473
4~8      1076
gt8       349
Name: count, dtype: int64
CoffeeHouse
less1    3385
1~3      3225
never    3179
4~8      1784
gt8      1111
Name: count, dtype: int64
CarryAway
1~3      4672
4~8      4258
less1    1856
gt8      1594
never     304
Name: count, dtype: int64
RestaurantLessThan20
1~3      5376
4~8      3580
less1    2093
gt8      1285
never     350
Name: count, dtype: int64
Restaurant20To50
less1    6077
1~3      3290
never    2325
4~8       728
gt8       264
Name: count, dtype: int64


The other 5 columns that contain missing data are structured similarly, with values ranging from 'never' to 'gt8'. The entries that contain missing data can either be dropped, or filled in with an assumption. For this study, empty cells will be treated as 'never', and filled in as such.

In [33]:
cols_with_missing = ['Bar', 'CoffeeHouse', 'CarryAway', 'RestaurantLessThan20', 'Restaurant20To50']
data[cols_with_missing] = data[cols_with_missing].fillna('never')
data.isnull().sum()

destination             0
passanger               0
weather                 0
temperature             0
time                    0
coupon                  0
expiration              0
gender                  0
age                     0
maritalStatus           0
has_children            0
education               0
occupation              0
income                  0
Bar                     0
CoffeeHouse             0
CarryAway               0
RestaurantLessThan20    0
Restaurant20To50        0
toCoupon_GEQ5min        0
toCoupon_GEQ15min       0
toCoupon_GEQ25min       0
direction_same          0
direction_opp           0
Y                       0
dtype: int64

There are no more empty cells. Viewing a sample of the data shows that the data is ready for analysis.

In [34]:
data.sample(10)

,destination,passanger,weather,temperature,time,coupon,expiration,gender,age,maritalStatus,...,CoffeeHouse,CarryAway,RestaurantLessThan20,Restaurant20To50,toCoupon_GEQ5min,toCoupon_GEQ15min,toCoupon_GEQ25min,direction_same,direction_opp,Y
11055,Work,Alone,Snowy,30,7AM,Restaurant(<20),1d,Male,46,Single,...,1~3,4~8,4~8,4~8,1,1,0,0,1,1
644,No Urgent Place,Friend(s),Sunny,55,2PM,Carry out & Take away,1d,Female,41,Single,...,1~3,1~3,1~3,never,1,1,0,0,1,1
2055,No Urgent Place,Friend(s),Sunny,80,2PM,Coffee House,1d,Female,36,Married partner,...,gt8,4~8,4~8,1~3,1,1,0,0,1,1
9721,Home,Alone,Snowy,30,10PM,Restaurant(<20),2h,Female,46,Single,...,1~3,gt8,4~8,1~3,1,1,1,0,1,0
9362,No Urgent Place,Friend(s),Sunny,30,10PM,Carry out & Take away,1d,Male,below21,Single,...,never,never,less1,less1,1,1,0,0,1,1
12256,Home,Partner,Rainy,55,6PM,Carry out & Take away,1d,Male,50plus,Married partner,...,1~3,4~8,1~3,1~3,1,0,0,1,0,1
2391,No Urgent Place,Friend(s),Sunny,80,6PM,Restaurant(<20),2h,Female,50plus,Married partner,...,never,4~8,gt8,less1,1,1,0,0,1,1
7824,No Urgent Place,Friend(s),Rainy,55,10PM,Bar,2h,Male,21,Single,...,4~8,1~3,less1,less1,1,1,0,0,1,1
10999,No Urgent Place,Alone,Sunny,80,10AM,Restaurant(<20),2h,Female,36,Single,...,less1,1~3,less1,less1,1,0,0,0,1,1
4474,Work,Alone,Sunny,80,7AM,Carry out & Take away,1d,Female,36,Single,...,never,4~8,1~3,less1,1,1,1,0,1,1
